Blog post; 1.37.0 announcement (#397)
rust-lang/blog.rust-lang.org: posts/2019-07-15-Rust-1.37.0.md

    match x {
        ByteOption::Some(y) => y + 1,
        ByteOption::None => 0,
    }
}
```

In implementations, `Self` acts like a type alias. So in Rust 1.37.0, you can also refer to `enum` variants with `Self::Variant`:

```rust
impl Coin {
    fn value_in_cents(&self) -> u8 {
        match self {
            Self::Penny => 1,
            Self::Nickel => 5,
            Self::Dime => 10,
            Self::Quarter => 25,
        }
    }
}
```

[type_rel_report]: https://github.com/rust-lang/rust/pull/61682/#issuecomment-502472847

To be more exact, Rust now allows you to refer to `enum` variants through *"type-relative resolution"*, `<MyType<..>>::Variant`. More details are available in [the stabilization report][type_rel_report].

### Built-in Cargo support for vendored dependencies

[vendor-crate]: https://crates.io/crates/cargo-vendor

After being available [as a separate crate][vendor-crate] for years, the `cargo vendor` command is now integrated directly into Cargo. The command fetches all your project's dependencies unpacking them into the `vendor/` directory, and shows the configuration snippet required to use the vendored code during builds.

> There are multiple cases where `cargo vendor` is already used in production: the Rust compiler `rustc` uses it to ship all its dependencies in release tarballs, and projects with monorepos use it to commit the dependencies' code in source control.

Reviewer: ```suggestion
There are multiple cases where `cargo vendor` is already used in production: The Rust compiler `rustc` uses it to ship all its dependencies in release tarballs, and projects with monorepos use it to commit the dependencies' code in source control.
```

Author: It's not a new sentence, so why capitalize? (If you used `.` as punctuation I'd be inclined to agree.)

Reviewer: Ah yeah, also fine by me.

### Using unnamed `const` items for macros

[unnamed_const_pr]: https://github.com/rust-lang/rust/pull/61347/

You can now create [unnamed `const` items][unnamed_const_pr]. Instead of giving your constant an explicit name, simply name it `_` instead. For example, in the `rustc` compiler we find:

```rust
/// Type size assertion where the first parameter
/// is a type and the second is the expected size.
#[macro_export]
macro_rules! static_assert_size {
    ($ty:ty, $size:expr) => {
        const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
        //    ^ Note the underscore here.
    }
}

static_assert_size!(Option<Box<String>>, 8); // 1.
static_assert_size!(usize, 8); // 2.
```

> Notice the second `static_assert_size!(..)`: thanks to the use of unnamed constants, you can define new items without naming conflicts. Previously you would have needed to write `static_assert_size!(MY_DUMMY_IDENTIFIER, usize, 8);`. Instead, with Rust 1.37.0, it now becomes easier to create ergonomic and reusable declarative and procedural macros for static analysis purposes.

Reviewer 2: Could anyone comment on why this is not solved by hygiene? E.g., if you wrote `const MY_DUMMY_IDENTIFIER: ...` in the body of the macro, I would expect hygiene to prevent naming conflicts

Author: What hygiene? =P

```rust
macro_rules! foo {
    () => {
        const MY_DUMMY_IDENTIFIER: () = ();
    }
}

foo!();
//foo!(); -- comment out ==> error.

fn main() {
    let _ = MY_DUMMY_IDENTIFIER; // OK
}
```

Macros 2.0 are still not a thing. ;)

Reviewer 2: Ah, right, `const`s are items not variables so they don't get hygiene :-(

> ### Profile-guided optimization

Author: r? @alexcrichton

### `#[repr(align(N))]` on `enum`s

[enum_align_pr]: https://github.com/rust-lang/rust/pull/61229
[ref_align_mod]: https://doc.rust-lang.org/reference/type-layout.html#the-alignment-modifiers
[ref_align_explain]: https://doc.rust-lang.org/reference/type-layout.html#size-and-alignment

[The `#[repr(align(N))]` attribute][ref_align_mod] can be used to raise the [alignment][ref_align_explain] of a type definition. Previously, the attribute was only allowed on `struct`s and `union`s. With Rust 1.37.0, the attribute can now also be used [on `enum` definitions][enum_align_pr]. For example, the following type `Align16` would, as expected, report `16` as the alignment whereas the natural alignment without `#[repr(align(16))]` would be `4`:

```rust
#[repr(align(16))]
enum Align16 {
    Foo { foo: u32 },
    Bar { bar: u32 },
}
```

The semantics of using `#[repr(align(N))` on an `enum` is the same as defining a wrapper struct `AlignN<T>` with that alignment and then using `AlignN<MyEnum>`:

```rust
#[repr(align(N))]
struct AlignN<T>(T);
```

### Library changes

[`BufReader::buffer`]: https://doc.rust-lang.org/std/io/struct.BufReader.html#method.buffer

> [`BufWriter::buffer`]: https://doc.rust-lang.org/std/io/struct.BufReader.html#method.buffer

Reviewer: ```suggestion
[`BufWriter::buffer`]: https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.buffer
```

[`Cell::from_mut`]: https://doc.rust-lang.org/std/cell/struct.Cell.html#method.from_mut
[`Cell::as_slice_of_cells`]: https://doc.rust-lang.org/std/cell/struct.Cell.html#method.as_slice_of_cells
[`DoubleEndedIterator::nth_back`]: https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html#method.nth_back
[`Option::xor`]: https://doc.rust-lang.org/std/option/enum.Option.html#method.xor
[`Wrapping::reverse_bits`]: https://doc.rust-lang.org/std/num/struct.Wrapping.html#method.reverse_bits
[`{i,u}{8,16,64,128,size}::reverse_bits`]: https://doc.rust-lang.org/std/primitive.u8.html#method.reverse_bits
[`slice::copy_within`]: https://doc.rust-lang.org/std/primitive.slice.html#method.copy_within

> In Rust 1.37.0 there have been a number of standard library stabilization:

Reviewer: ```suggestion
In Rust 1.37.0 there have been a number of standard library stabilizations:
```

- [`BufReader::buffer`] and [`BufWriter::buffer`]
- [`Cell::from_mut`]
- [`Cell::as_slice_of_cells`]
- [`DoubleEndedIterator::nth_back`]
- [`Option::xor`]
- [`{i,u}{8,16,64,128,size}::reverse_bits`] and [`Wrapping::reverse_bits`]
- [`slice::copy_within`]

### Other changes

[relnotes-cargo]: https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#cargo-137-2019-08-15
[relnotes-clippy]: https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md#rust-137

> There are other changes in the Rust 1.37 release: check out what changed in [Rust][notes], [Cargo][relnotes-cargo], and [Clippy][relnotes-clippy].

Reviewer: ```suggestion
There are other changes in the Rust 1.37 release: Check out what changed in [Rust][notes], [Cargo][relnotes-cargo], and [Clippy][relnotes-clippy].
```

## Contributors to 1.37.0

Many people came together to create Rust 1.37.0. We couldn't have done it
without all of you. [Thanks!](https://thanks.rust-lang.org/rust/1.37.0/)

## New sponsors of Rust infrastructure

> Along with the release, we’d like to thank two new sponsors of Rust’s infrastructure who provided the resources needed to make Rust 1.37.0 happen: Amazon Web Services (AWS) and Microsoft.

Reviewer: ```suggestion
Along with the release, we'd like to thank two new sponsors of Rust's infrastructure who provided the resources needed to make Rust 1.37.0 happen: Amazon Web Services (AWS) and Microsoft.
```

> Along with the release, we'd like to thank two new sponsors of Rust's infrastructure who provided the resources needed to make Rust 1.37.0 happen: Amazon Web Services (AWS) and Microsoft.

Reviewer 2: ```suggestion
We'd like to thank two new sponsors of Rust's infrastructure who provided the resources needed to make Rust 1.37.0 happen: Amazon Web Services (AWS) and Microsoft.
```

> - AWS has generously covered our bill for hosting compilers on S3, serving those compilers to users through CloudFront, preventing regressions with Crater on EC2, and managing numerous other Rust-related infrastructure hosted on AWS. 

Reviewer 2: ```suggestion
- AWS has provided hosting for release artifacts (compilers, libraries, tools, and source code), serving those artifacts to users through CloudFront, preventing regressions with Crater on EC2, and managing other Rust-related infrastructure hosted on AWS. 
```

> - Microsoft has generously sponsored builders for Rust’s CI infrastructure, notably the extremely resource intensive rust-lang/rust repository.

Reviewer 2: ```suggestion
- Microsoft has sponsored builders for Rust’s CI infrastructure, notably the extremely resource intensive rust-lang/rust repository.
```
